src/main/java/org/broadinstitute/hellbender/tools/spark/utils/HopscotchCollection.java

```diff
@@ -562,6 +557,25 @@ public T next() {
             final T result = buckets[currentIndex];
             removeIndex = currentIndex;
             removePrevIndex = prevIndex;
+            advanceUntilEquivalent();
+            return result;
+        }
+
+        @Override
+        public void remove() {
+            super.remove();
+            if ( currentIndex != NO_ELEMENT_INDEX ) {
+                checkEquivalence();
+            }
+        }
+
+        private void checkEquivalence() {
```

Comment:
Can't eliminate either method without doing a redundant equivalence test.
ensureEquivalence is effectively: ```while ( !equivalent ) advance;```
advanceUntilEquivalent is: ```do advance; while ( !equivalent );```
If I called ensureEquivalence in the next() method rather than advanceUntilEquivalent I'd be doing an extra equivalence test on the current key, which is already known to be equivalent.
Maybe an extra key compare is trivial, but maybe it's not -- you'd need to know something about the key.
I did the rename to ensureEquivalence, which does seem like a better name.